### PR TITLE
Multi hop support

### PR DESCRIPTION
**Pull request by ethanfrey**
Builds on #11 (merge that first)

- [x] Implement and test multi-hop for exact input
- [x] Implement and test multi-hop for exact output

### CODE REVIEW

File: app/wasm/message_plugin.go [hunk 1]
```diff
@@ -107,9 +108,17 @@ func performSwap(keeper *gammkeeper.Keeper, ctx sdk.Context, contractAddr sdk.Ac
 			PoolId:       swap.First.PoolId,
 			TokenInDenom: swap.First.DenomIn,
 		}}
+		output := swap.First.DenomOut
+		for _, step := range swap.Route {
+			routes = append(routes, gammtypes.SwapAmountOutRoute{
+				PoolId:       step.PoolId,
+				TokenInDenom: output,
+			})
+			output = step.DenomOut
```
Comment (maurolacy) [inline app/wasm/message_plugin.go:117]: Ah, now I fully understand your idea.

Makes sense. My only concern would be, if this is fully clear / can be fully derived, or assumed, from the `ExactOut` specs.

**Review by maurolacy — COMMENTED** [review]
Looks good. Will review tests later.

File: app/wasm/test/custom_msg_test.go [hunk 2]
```diff
@@ -209,6 +209,163 @@ func TestSwapMsg(t *testing.T) {
...
+
+		"exact in: 2 step multi-hop": {
+			msg: func(state BaseState) *wasmbindings.SwapMsg {
+				return &wasmbindings.SwapMsg{
+					First: wasmbindings.Swap{
+						PoolId:   state.StarPool,
+						DenomIn:  "ustar",
+						DenomOut: "uosmo",
+					},
+					// Note: you must use empty array, not nil, for valid Rust JSON
```
Comment (maurolacy) [inline app/wasm/test/custom_msg_test.go]: The Note is now outdated / irrelevant?

File: app/wasm/test/custom_msg_test.go [hunk 3]
```diff
@@ -209,6 +209,163 @@ func TestSwapMsg(t *testing.T) {
...
+					Route: []wasmbindings.Step{{
```
Comment (maurolacy) [inline app/wasm/test/custom_msg_test.go:221]: Nice "double bracket" construct.

File: app/wasm/test/custom_msg_test.go [hunk 4]
```diff
@@ -209,6 +209,163 @@ func TestSwapMsg(t *testing.T) {
...
+							Input:     sdk.NewInt(240000000),
+							MinOutput: sdk.NewInt(1999000),
+						},
+					},
+				}
+			},
+			initFunds: sdk.NewInt64Coin("ustar", 240000000),
+			finalFunds: []sdk.Coin{
+				// 240 STAR -> 6 OSMO
+				// 6 OSMO -> 2 ATOM (with minor rounding)
```
Comment (maurolacy) [inline app/wasm/test/custom_msg_test.go:236]: It's clear to me now: The exact in / out sequence always define the same: input -> intermediate outputs -> final output.

And we handle the differences between them under the hood. I agree this is a better design / API than the Osmosis one. Was just confused / doubting about the exact meaning of the difference between them.

**Review by maurolacy — APPROVED** [review]
LGTM.

File: app/wasm/message_plugin.go (hunk 1 above)
Comment (ethanfrey) [reply app/wasm/message_plugin.go:117]: This is equivalent to the Osmosis API in that there is a 1:1 mapping of all valid inputs.
This one just seems more readable, consistent to me, in that you can switch between fixing input and fixing output very easily.

File: app/wasm/test/custom_msg_test.go (hunk 2 above)
Comment (ethanfrey) [reply app/wasm/test/custom_msg_test.go]: Good point, there are 4 places this note was no longer needed (cut and pasted).
Removing them

**Pull request merged**